Add S3RemoteStorageManager

ivanyu opened the pull request.

ivanyu committed:
Add S3RemoteStorageManager

satishd reviewed (commented):
thanks @ivanyu for the PR. Overall LGTM except for the below comment.

`RemoteLogContext` can store the protocol version of how the key name is computed from `RemoteLogMetadata` instead of having all the file names for that segment. It can be just a version number(maybe short) of the protocol on how it is done. If we ever change the key naming convention you can update the version. S3RemoteStorageManager should handle all the versions that it supports. This will reduce the amount of data duplication stored in RLMM for each entry. 

In your case, version 0 can eb of the form: 
key-name is: `<topic-partition>/<remote-log-segment-uuid>.<suffix>`

suffix can be: `.log` or `.timeindex` or `.offsetindex`

Hangleton commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,215 @@
+        Objects.requireNonNull(logSegmentData, "logSegmentData must not be null");
+
+        try {
+            final String logFileKey = s3Key(remoteLogSegmentId, logSegmentData.logSegment().getName());
```
Fail-over mode/atomicity: in case of failures of a given upload - say, the time index - I think removing the files already upload would be beneficial. This guarantees that no segment and associated metadata is partially uploaded. Arguably, the remote log metadata manager in Kafka will identify the failure and put the metadata in a state reflecting it / or remove the entry for this upload altogether.

ivanyu replied:
Since each upload is uniquely identified with a UUID, there should be no partial upload visible to RLM.

So it seems that deleting the partial uploads is mostly about garbage collection. In this regard, we can offer only the best effort deletion. For example, if S3 becomes inaccessible for long, we can't block in the method waiting it to come back. I'll make it attempt the clean-up deletion once.

Hangleton commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,215 @@
+                logSegmentData.offsetIndex().getName(),
+                logSegmentData.timeIndex().getName()
+            );
+        } catch (final SdkClientException e) {
```
Also catch any exception so as to make sure all exceptions propagated by this method are of type `RemoteStorageException`, that helps caller dealing with failure modes as a single type of exception can be expected. Additionally, contextual information can be added to the `RemoteStorageException` in the future to better categorize the type of error (e.g. invalid configuration, unavailability of the remote storage, etc.), again helping with failure handling in the RSM. To that end, the categories of errors offered to implementations should be defined in the framework itself and not be storage-implementation specific. The example of HDFS and S3 storage allows to collect types failures from which the least common denominators can be extracted.

Hangleton commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,215 @@
+
+    @Override
+    public void close() throws IOException {
+        // intentionally left blank
```
Don't we need to close S3 client (not sure it is offered by the API?)

ivanyu replied:
Yes, sure

Hangleton reviewed (commented):
Thanks for the PR. Very readable. Left couple of comments for the storage manager. That involves both implementation and framework level attributes.

ivanyu committed:
Shut down TransferManager in S3RemoteStorageManager on close

ivanyu committed:
Catch all exceptions and attempt to clean up

This makes S3RemoteStorageManager to catch all exceptions on the interaction with S3 and propagate them further in wrapped in RemoteStorageException.

Also, it makes one attemp to clean up partially uploads.

ivanyu committed:
Make S3RemoteLogSegmentContext versioned and store only offset

`S3RemoteLogSegmentContext` is now versioned.

Also, `S3RemoteLogSegmentContext` now contains only the base offset for the upload instead of full file names to save space.

ivanyu commented:
@satishd @Hangleton thank you for the comments, guys. I made some changes, please have a look when you have a minute.

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteLogSegmentContext.java:
```diff
@@ -24,59 +24,52 @@
     public static S3RemoteLogSegmentContext fromBytes(final byte[] bytes) {
-        final Struct struct = SCHEMA.read(ByteBuffer.wrap(bytes));
+        // When there are more schema versions, read conditionally.
+        final Struct struct = SCHEMA_V0.read(ByteBuffer.wrap(bytes));
```
I prefer version outside of the payload schema. 
- read version, which can be short 
- get the schema for that version and pass the remaining bytes to building the required structure. 

WDYT?

ivanyu replied:
Yeah, I though this to be done when version 1 appear as it should be compatible on the byte level. But you're right, better to do this way from the beginning. Done.

satishd reviewed (commented):
thanks @ivanyu for the update, left a comment.

Hangleton commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,241 @@
+        Objects.requireNonNull(remoteLogSegmentId, "remoteLogSegmentId must not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData must not be null");
+
+        final long baseOffset = Log.offsetFromFileName(logSegmentData.logSegment().getName());
```
Since this implementation will be outside of the Apache Kafka codebase eventually, will it have access to `Log`?

ivanyu replied:
Of course we can add the runtime dependency on `core`, but the dependency surface is so limited that it's easier to just copy the implementation of these couple of functions. Done.

Hangleton replied:
Thanks.

ivanyu committed:
Separate version in S3RemoteLogSegmentContext schema

ivanyu committed:
Remove dependency on kafka.log.Log

ivanyu commented:
@satishd @Hangleton addressed your comments, guys

satishd reviewed (approved):
thanks @ivanyu for addressing the review comments, LGTM.

Hangleton reviewed (commented):
Thanks for the PR. Next step is to try with S3.
